src/io_uring.c

```diff
@@ -0,0 +1,92 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+static size_t io_uring_queue_len = 0;
+
+void initIOUring(void) {
+    if (server.io_uring_enabled) {
+        struct io_uring_params params;
+        struct io_uring *ring = zmalloc(sizeof(struct io_uring));
+        memset(&params, 0, sizeof(params));
+        /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+         * point to the shared memory containing the io_uring queues.
+         * On failure -errno is returned. */
+        int ret = io_uring_queue_init_params(IO_URING_DEPTH, ring, &params);
+        if (ret != 0) {
+            /* Warning if user enable the io_uring in config but system doesn't support yet. */
+            serverLog(LL_WARNING, "Failed to initialize io_uring: %s", strerror(-ret));
+            zfree(ring);
+            server.io_uring = NULL;
+        } else {
+            serverLog(LL_NOTICE, "io_uring enabled.");
+            server.io_uring = ring;
+        }
+    }
+}
+
+int canFsyncUsingIOUring(void) {
+    return server.aof_state == AOF_ON && server.aof_fsync == AOF_FSYNC_ALWAYS && server.io_uring_enabled &&
+           server.io_uring;
+}
+
+/* Submit fdatasync request to io_uring. */
+void ioUringPrepFsyncAndSubmit(int fd) {
```

Comment:
Similarly, I would suggest passing the opaque io_uring pointer in
```suggestion
void ioUringPrepFsyncAndSubmit(void *io_uring, int fd) {
```

Reply:
Using `void *` to make structs opaque is not a good method for type safety. It throws away type information.

It's better to declare the struct without defining it, like `struct io_uring;`. Then it can be used for pointers but it's not possible to access the fields inside it.

In `io_uring.h` we can declare `struct io_uring;`, without including `<iouring.h>`.

In `io_uring.c` we can include `<iouring.h>` and get the full definition of `struct io_uring` which we need for e.g. `zmalloc(sizeof(struct io_uring))`.

Reply:
Good points, @zuiderkwast. +1 on `struct io_uring` for information hiding.